Enable complex autograd for torch.unfold

iramazanli opened the pull request:
Fixes  #51875

Commit:
enable complex autograd for unfold

Commit:
add test cases for complex supporting unfold

Commit:
updates to the opinfo unfold test

nikitaved committed:
Stable sort for the CPU take 2. (#51790)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/38681.
A duplicate of https://github.com/pytorch/pytorch/pull/50052 created to become importable to the fb internal tests.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/51790

Reviewed By: agolynski

Differential Revision: D26279045

Pulled By: glaringlee

fbshipit-source-id: 348e171dee9c370a76002b65d0c82c329f57a421

H-Huang committed:
Change TCPStore world_size and is_master to be optional (#51809)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/51809

Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

Test Plan: Imported from OSS

Reviewed By: heitorschueroff

Differential Revision: D26461770

Pulled By: H-Huang

fbshipit-source-id: 5b2157029c73e8706e158cd49ecce60c9f3a7f41

suo committed:
[package] Introduce Importer to manage module namespace collisions. (#51975)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/51975

See comments in code.

Test Plan: Imported from OSS

Reviewed By: zdevito

Differential Revision: D26340592

Pulled By: suo

fbshipit-source-id: 61b16bafad15e19060710ad2d8487c776d672847

suo committed:
[package] rename ex/importer.py to package_ex/importer.py (#52320)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52320

as title

Test Plan: Imported from OSS

Reviewed By: zdevito

Differential Revision: D26468416

Pulled By: suo

fbshipit-source-id: 890eecea76426918daff900402fbcbc149e48535

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1309,6 +1309,40 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+
+    sample_inputs = []
+    for shape, arguments in test_cases:
+        sample_inputs += [SampleInput(make_tensor(dim, device, dtype,
```
`sample_inputs += [SampleInput(make_tensor(shape, device, dtype,`

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2290,6 +2324,11 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
            sample_inputs_func=sample_inputs_hstack_dstack_vstack),
+    OpInfo('unfold',
+           dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
+#           test_inplace_grad=False,
```
you probably need to uncomment both these lines

Commit:
Trace linear as aten::linear (#51897)

Summary:
https://github.com/pytorch/pytorch/pull/51613 made `torch.nn.functional.linear` compile as `aten::linear`, extend the same behavior with tracing.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/51897

Reviewed By: albanD

Differential Revision: D26320711

Pulled By: eellison

fbshipit-source-id: a26d3c37323a0706313c6ebb210bad60eec6a64b

Commit:
suggested changes

malfet committed:
Add explicit cudart_static dependency for cublas_static (#52509)

Summary:
Fixes following error during static linking, by enforcing that cudart dependency is put after cublasLt
```
/usr/bin/ld: /usr/local/cuda/lib64/libcublasLt_static.a(libcublasLt_static.a.o): undefined reference to symbol 'cudaStreamWaitEvent@libcudart.so.11.0'
/usr/local/cuda/lib64/libcudart.so: error adding symbols: DSO missing from command line
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/52509

Reviewed By: janeyx99

Differential Revision: D26547622

Pulled By: malfet

fbshipit-source-id: 4e17f18cf0ab5479a549299faf2583a79fbda4b9

shubhambhokare1 committed:
[ONNX] Update fuseLogSoftmaxNllLoss function to handle autocasting (#51729) (#52349)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52349

Adds a check for patterns for cases with autocasting enabled in which a cast node is inserted before the NegativeLogLikelihoodLoss
node and causing these patterns below not to be recognizable by peephole pass function

Test Plan: Imported from OSS

Reviewed By: malfet

Differential Revision: D26490326

Pulled By: SplitInfinity

fbshipit-source-id: 4a6d806acc51b4696fd3932734d55af075fba6b1

shubhambhokare1 committed:
[ONNX] Update LayerNorm symbolic to handle autocasting (#52199) (#52350)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52350

When onnx export creates a 0-dim tensor of constant type, this action overrides the type promotion logic as quoted in #9515. In order to prevent this from happening this PR adds the following functionality.
If the data type is a floating point type, it is converted to a 0-dim double tensor, else it is converted to a 0-dim tensor of its original type

Test Plan: Imported from OSS

Reviewed By: malfet

Differential Revision: D26490325

Pulled By: SplitInfinity

fbshipit-source-id: 4c47c69c9b6523d2e45b74c2541d6d8ca7e28fc9

neginraoof committed:
[ONNX] Fix for sequence of mutations in blocks (#51577) (#52347)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52347

Fixes consecutive mutations in a tensor inside blocks.
Also, support append and pop in blocks.

Test Plan: Imported from OSS

Reviewed By: malfet

Differential Revision: D26490328

Pulled By: SplitInfinity

fbshipit-source-id: f0cdc706d2793e1f4eb0503d3e0f63f4127ea47a

jerryzh168 committed:
[quant][graphmode][fx] Enable test for non quantized input for add/mul (#52412)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52412

When the input is not quantized, we'll still quantize add/mul

Test Plan: Imported from OSS

Reviewed By: supriyar

Differential Revision: D26503347

fbshipit-source-id: 457b3444c50e5b49b911b04c67684f5eead78ec9

Commit:
[JIT] Display an error message when with item is not an object (#52335)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52335

**Summary**
`with` statements can only be used with objects that have `__enter__`
and `__exit__` defined. At present, any attempt to use an expression
that returns something that is not an instance of a class type results
in a cryptic internal assert failure instead of a useful error message.
This is because the code that generates IR for `with` statements uses
`Type::expect` as if it were `Type::cast`; that is, as if it returns
`nullptr` on failure.

This commit fixes this issue by checking the `kind()` of the type of the
expression used as the with item before calling `expect<ClassType>()` on
it.

**Test Plan**
This commit adds a unit test to `test_with_errors` to test this case.

Test Plan: Imported from OSS

Reviewed By: gmagogsfm

Differential Revision: D26504909

Pulled By: SplitInfinity

fbshipit-source-id: 92d108e0c010370fd45131a57120f50c0b85c401

Commit:
[JIT] Allow __exit__ to have a return value (#52336)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52336

**Summary**
In Python, the boolean interpretation of the return value of `__exit__` of objects that are used as context managers with `with` statements is used to determine
whether or not to propagate exceptions thrown inside the body of the with
statement. This latter feature is not possible to add to TorchScript at
the moment, but the current requirement that `__exit__` not have any
return values can make it difficult to script a context manager whose
`__exit__` *does* have a return value.

Accordingly, this commit removes the requirement that `__exit__` must
not have any return value. TorchScript does not interpret this return
value in the same way Python does (or at all), but this should make it
easier to share context managers between eager mode and script code.

**Test Plan**
This commit adds a return value to one of the context managers used in
`TestWith`.

Test Plan: Imported from OSS

Reviewed By: gmagogsfm

Differential Revision: D26504910

Pulled By: SplitInfinity

fbshipit-source-id: 2ab635a24d111ac25df4e361b716be8fada5128e

ljk53 committed:
[pytorch] reintroduce static dispatch (#51957)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/51957

This is a simplified version of #51554.

Compared to #51554, this version only supports statically dispatching to
a specific backend. The benefit is that it skipped the dispatch key
computation logic thus has less framework overhead. The downside is that
if input tensors do not match the specified backend it will throw error
instead of falling back to regular dispatch.

Sample code:
```
Tensor empty(IntArrayRef size, TensorOptions options, c10::optional<MemoryFormat> memory_format) {
    return at::cpu::empty(size, options, memory_format);
}

// aten::conj(Tensor(a) self) -> Tensor(a)
Tensor conj(const Tensor & self) {
    return at::math::conj(self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_out(Tensor & out, const Tensor & self) {
    return at::cpu::conj_out(out, self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_outf(const Tensor & self, Tensor & out) {
    return at::cpu::conj_out(out, self);
}

// aten::_conj(Tensor self) -> Tensor
Tensor _conj(const Tensor & self) {
    return at::defaultbackend::_conj(self);
}
```

For ops without the specific backend dispatch, it will throw error:
```
// aten::_use_cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank) -> bool
bool _use_cudnn_ctc_loss(const Tensor & log_probs, const Tensor & targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t blank) {
    TORCH_CHECK(false, "Static dispatch does not support _use_cudnn_ctc_loss for CPU.");
}
```

Differential Revision: D26337857

Test Plan: Imported from OSS

Reviewed By: bhosmer

Pulled By: ljk53

fbshipit-source-id: a8e95799115c349de3c09f04a26b01d21a679364

CarlosJose126 committed:
Example LSTMCell (#51983)

Summary:
Fixes #{51801}
LSTMCell example updated

Pull Request resolved: https://github.com/pytorch/pytorch/pull/51983

Reviewed By: agolynski

Differential Revision: D26467104

Pulled By: zou3519

fbshipit-source-id: 31c8bf89b21cd2f748b2cc28a74169082d81503c

navahgar committed:
Adding a flag to enable CPU fusion in benchmarks (#48612)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/48612

Test Plan: python -m benchmarks.tensorexpr --device cpu --mode fwd --jit_mode trace --cpu_fusion element

Reviewed By: heitorschueroff

Differential Revision: D26548643

Pulled By: navahgar

fbshipit-source-id: adb537818d77c9b6b0fe434ae6d963a5f348ad24

peterjc123 committed:
Automatically set BUILD_SPLIT_CUDA for cpp exts (#52503)

Summary:
Fixes https://github.com/pytorch/vision/pull/3418#issuecomment-781673110

Pull Request resolved: https://github.com/pytorch/pytorch/pull/52503

Reviewed By: malfet

Differential Revision: D26546857

Pulled By: janeyx99

fbshipit-source-id: a100b408e7cd28695145a1dda7f2fa081bb7f21f

kshitij12345 committed:
[reland][complex] `masked_fill`: Complex Autograd support and update masked_scatter skips. (#52483)

Summary:
Reland https://github.com/pytorch/pytorch/issues/52035

Pull Request resolved: https://github.com/pytorch/pytorch/pull/52483

Reviewed By: heitorschueroff

Differential Revision: D26545097

Pulled By: anjali411

fbshipit-source-id: f154c239183279be381a7393a8226778b36148bb

swolchok committed:
[PyTorch] Sync TORCH_INTERNAL_ASSERT optis with TORCH_CHECK (#52226)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52226

This gets TORCH_INTERNAL_ASSERT to parity with TORCH_CHECK in terms of optimization for 0 or 1 argument.
ghstack-source-id: 121877054

(Note: this ignores all push blocking failures!)

Test Plan:
Compare generated assembly for
```
#include <c10/util/Exception.h>

void f(bool b) {
  TORCH_INTERNAL_ASSERT(b, "message");
}

void g(bool b) {
  TORCH_INTERNAL_ASSERT(b);
}

void h(bool b) {
  TORCH_INTERNAL_ASSERT(b, "message", random());
}
```

before/after this diff.
Before: P174916324
After: P174916411

Before, f and g called out to outlined lambdas to build
std::strings. After, they load string constants and call
torchInternalAssertFail. Similarly, h calls random() and c10::detail::_str_wrapper() inline and then calls out to torchInternalAssertFail. As with D26380783 (https://github.com/pytorch/pytorch/commit/efbb854ed8a9df35c0ea896c2d216ab3da10d677), I hope to solve the problem of outlining the random & _str_wrapper calls separately.

Profile AdIndexer benchmark & verify that toTensor() is still inlined (it calls TORCH_INTERNAL_ASSERT with an integer argument, like `h` above).

Reviewed By: bhosmer

Differential Revision: D26410575

fbshipit-source-id: f82ffec8d302c9a51f7a82c65bc698fab01e1765

swolchok committed:
[c10] Take advantage of c10::str optis for simple CAFFE_ENFORCE (#52223)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52223

After the previous diffs, `c10::str()` will return a
`CompileTimeEmptyString` when passed 0 arguments and a `const char*` when
passed 1 `const char *` argument. We can take advantage of this to
outline further std::string creation from CAFFE_ENFORCE.
ghstack-source-id: 121877053

(Note: this ignores all push blocking failures!)

Test Plan:
Compare assembly for
```
#include <c10/util/Logging.h>

void f(bool b) {
  CAFFE_ENFORCE(b);
}

void g(bool b) {
  CAFFE_ENFORCE(b, "message");
}

void h(bool b) {
  CAFFE_ENFORCE(b, "message", random());
}
```

before & after this diff.

before: P174902847
after: P174902912

f & g are clearly much improved, and h is about the same.

(I tried measuring caffe2 perf on the AdIndexer MergeNet benchmark, but didn't see a win, which makes sense because the change is small.)

Reviewed By: bhosmer

Differential Revision: D26405181

fbshipit-source-id: c51a9e459ae7d9876494a83ade6f6fe725619512

swolchok committed:
[PyTorch] Make c10::str(const char*) return const char* (#52222)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52222

`c10::str()` is often used with variadic macros. It can be more efficient to get a C string out if you put a C string in, like if you are able to defer std::string creation to an outlined function or even never do it at all. Meanwhile, there is an implicit conversion from const char* to std::string, so users who expected a std::string will still make one.
ghstack-source-id: 121877052

(Note: this ignores all push blocking failures!)

Test Plan: CI

Reviewed By: bhosmer

Differential Revision: D26419663

fbshipit-source-id: 400bef71e6a0004b5914f5f511ea0e04e0d7599b

anjali411 commented:
@iramazanli there seems to be a merge conflict in `tools/autograd/gen_variable_type.py`. could you please rebase?

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2298,7 +2337,7 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
            skips=(
                SkipInfo('TestCommon', 'test_variant_consistency_jit',
-                        dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16)),
+                        dtypes=all_types_and_complex_and(torch.bool, torch.fl16, torch.bfloat16)),
```
please undo this change

jerryzh168 committed:
[quant][graphmode][fx] Guard the supported quantization type for add/mul (#52413)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52413

TODO: We'll need to add this guard for other ops as well

(Note: this ignores all push blocking failures!)

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_mul_add_fp16_config

Imported from OSS

Reviewed By: supriyar

Differential Revision: D26503348

fbshipit-source-id: 5aaba518742a516cc3521fd5f23f1a264d2973e2

bmon581 committed:
Change cmake to allow building with MLC kick-off build (#51326)

Summary:
- Allows build process to build with MLC enabled if subrepo folder mlc is in path and we can link against ML Compute on macOS BigSur
- To build with MLC enabled you will need to clone the mlc repo inside the pytorch repository.
- We need both this change and https://github.com/pytorch/pytorch/pull/50634 on pytorch/pytorch to enable the `mlc` device.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/51326

Reviewed By: glaringlee

Differential Revision: D26533138

Pulled By: malfet

fbshipit-source-id: 0baa06b4eb2d62dbfc0f6fc922096cb0db1cc7d1

anjali411 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -91,7 +91,7 @@
     'replication_pad1d', 'replication_pad2d', 'replication_pad3d',
     'replication_pad1d_backward', 'replication_pad2d_backward', 'replication_pad3d_backward',
-    'diag', 'masked_scatter', 'masked_select', 'masked_fill', 'index_fill', 'trace'
+    'diag', 'masked_scatter', 'masked_select', 'index_fill', 'trace', 'unfold'
```
could you add `masked_fill` back to this list?

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2353,6 +2387,11 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
            ),
            sample_inputs_func=sample_inputs_hstack_dstack_vstack),
+    OpInfo('unfold',
+           dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
```
you need `op=lambda x, args: x.unfold(*args),`

Commit:
fix by adding op for unfold

Commit:
final updates

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2353,9 +2387,15 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
     OpInfo('vstack',
            # gradcheck expects the input arguments as a flat list
-           op=lambda *args: torch.vstack([*args]),
+           op=lambda x, *args: x.unfold(*args),
```
@iramazanli could you please undo this change?

Commit:
fixing unfold and vstack confusion

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2353,6 +2387,12 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+    OpInfo('unfold',
+           op=lambda x, *args: x.unfold(*args),
+           dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
+           test_inplace_grad=False,
```
From the CI failures, it seems like `test_variant_consistency_jit` is failing. This should resolve the issue:
```
 skips=(
              # torch.unfold does not exist so we get a RuntimeError.
              SkipInfo('TestCommon', 'test_variant_consistency_jit',
                       dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16)),
          ),
```

anjali411 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -91,7 +91,7 @@
     'replication_pad1d', 'replication_pad2d', 'replication_pad3d',
     'replication_pad1d_backward', 'replication_pad2d_backward', 'replication_pad3d_backward',
-    'diag', 'masked_scatter', 'masked_select', 'masked_fill', 'index_fill', 'trace'
+    'diag', 'masked_scatter', 'masked_select', 'masked_fill', 'index_fill', 'trace', 'unfold'
```
add `unfold_backward` to this list as well

iramazanli committed:
fiing errors

anjali411 commented:
@iramazanli could you please rebase and resolve the merge conflict?
